src/network/connection.cpp

```diff
@@ -62,18 +62,28 @@ namespace con
 
 #define PING_TIMEOUT 5.0
 
-BufferedPacket makePacket(Address &address, const SharedBuffer<u8> &data,
+u16 BufferedPacket::getSeqnum() const
+{
+	if (data.size() < BASE_HEADER_SIZE + 3) {
+		//std::cout << (int)data[0xFFFFFFFFF] << std::endl;
+		throw ConnectionException("Packet too short to contain a seqnum");
+	}
+
+	return readU16(&data[BASE_HEADER_SIZE + 1]);
+}
+
+BufferedPacketPtr makePacket(Address &address, const SharedBuffer<u8> &data,
 		u32 protocol_id, session_t sender_peer_id, u8 channel)
 {
 	u32 packet_size = data.getSize() + BASE_HEADER_SIZE;
-	BufferedPacket p(packet_size);
-	p.address = address;
+	auto p = std::make_shared<BufferedPacket>(packet_size);
```

Comment:
Is it possible to use unique_ptr instead?

Reply:
I wish it were that easy. Packets are copied in `ReliablePacketBuffer::getTimedOuts`. Shared pointers can help a lot to deal with accidental copies to the stack (function arguments or by return value).

Reply:
Also are we sure that we are make copies here? Most compiler with do return value optimization and construct the return object in place. I guess you are more concerned later when we pass the buffer to other methods...?

Reply:
I am convinced that `std::move` could still copy data, like implied by cppreference:

> `std::move` is used to indicate that an object `t` may be "moved from", i.e. allowing the efficient transfer of resources from `t` to another object. 

Hence working with pointers - preferably smart pointers - is the preferred way. Either we implement it ourselves, or use `std::shared_pointer` to properly deal with function argument copies - whether they may happen intentionally or by accident. Copying a "pointer" is much cheaper and reliable than `std::move` with entire structs.